Translation of the accessibility page (#62)
reactjs/pt-PT.reactjs.org: content/docs/accessibility.md

permalink: docs/accessibility.html
---

## Porque Acessibilidade? {#why-accessibility}

> A acessibilidade na web (tambem conhecido por [**a11y**](https://en.wiktionary.org/wiki/a11y)) é o design e a criação de websites que podem ser usados por todos. O suporte à acessibilidade é necessário para permitir que as tecnologias assistivas interpretem as páginas da web.

Reviewer: ```suggestion
A acessibilidade na web (também conhecida por [**a11y**](https://en.wiktionary.org/wiki/a11y)) é o design e a criação de websites que podem ser usados por todos. O suporte à acessibilidade é necessário para permitir que as tecnologias assistivas interpretem as páginas da web.
```

Reviewer: Acentuação e concordância do gênero

Author: Thanks. ;)

/>
```

> ## Semântica HTML {#semantic-html}

Author: Concordo mais com "Semântica HTML"

> ## Semântica HTML {#semantic-html}

Reviewer: cool

> A semântica é a base de uma aplicação web acessível. Se usarmos corretamente os elementos HTML para reforçar o significado da informação colocada nos nossos websites, muitas das vezes a acessibilidade pode ser gratuita.

Author: o mesmo aqui mantinha a semântica apenas e adicionava apenas o "vir de forma" gratuita.

> A semântica é a base de uma aplicação web acessível. Se usarmos corretamente os elementos HTML para reforçar o significado da informação colocada nos nossos websites, muitas das vezes a acessibilidade pode ser gratuita.

Reviewer: alterado!

- [MDN HTML elements reference](https://developer.mozilla.org/en-US/docs/Web/HTML/Element)

> Às vezes, quebramos a semântica HTML para adicionar elementos `<div>` ao nosso JSX apenas para fazer funcionar no React, especialmente quando trabalharmos com listas (`<ol>`, `<ul>` e `<dl>`) e com `<table>`.

Author: Manitnha "semântica HTML" em vez de HTML semântica

> Às vezes, quebramos a semântica HTML para adicionar elementos `<div>` ao nosso JSX apenas para fazer funcionar no React, especialmente quando trabalharmos com listas (`<ol>`, `<ul>` e `<dl>`) e com `<table>`.

Reviewer: alterado

}
```

> Quando não é necessário passar propriedades para a tag Fragment podemos usar a [syntax curta](/docs/fragments.html#short-syntax), se a configuração suportar:

Reviewer: ```suggestion
Quando não é necessário passar props para a tag Fragment podemos usar a [síntaxe curta](/docs/fragments.html#short-syntax), se a configuração suportar:
```

}
```

> Para mais informação, veja a [documentação sobre Fragments](/docs/fragments.html).

Reviewer: ```suggestion
Para mais informação, vê a [documentação sobre Fragments](/docs/fragments.html).
```

Reviewer: Optamos [aqui](https://github.com/reactjs/pt-PT.reactjs.org/issues/9) por abordar o leitor por _tu_ e não _você_.

Embora estas práticas HTML padrão possam ser usadas diretamente no React, observa que o atributo `for` está escrito como `htmlFor` em JSX:

```javascript{1}
<label htmlFor="inputNome">Nome:</label>
<input id="inputNome" type="text" name="nome"/>
```

### Notificar erros ao utilizador {#notifying-the-user-of-errors}

Situações de erro precisam de ser entendidas por todos os utilizadores. Os artigos seguintes mostram como exportar os erros aos leitores de ecrã:

- [The W3C demonstrates user notifications](https://www.w3.org/WAI/tutorials/forms/notifications/)
- [WebAIM looks at form validation](https://webaim.org/techniques/formvalidation/)

## Controlo de focus {#focus-control}

Certifique-se que a aplicação web possa ser utilizada apenas com o teclado:

- [WebAIM talks about keyboard accessibility](https://webaim.org/techniques/keyboard/)

### Focus no teclado e focus no contorno {#keyboard-focus-and-focus-outline}

Focus no teclado refere-se ao elemento DOM que foi selecionado e que aceita ações do teclado. Podemos ver o contorno na imagem a seguir:

<img src="../images/docs/keyboard-focus.png" alt="Contorno azul à volta da ligação selecionada." />

> Apenas use CSS que elimine este contorno, por exemplo definindo `outline: 0`, se for substituir por outra implementação de focus.

Reviewer: ```suggestion
Usa apenas CSS que elimine este contorno, por exemplo definindo `outline: 0`, se for substituir por outra implementação de focus.
```

### Mecanismos para avançar para o conteúdo desejado {#mechanisms-to-skip-to-desired-content}

> Providencie mecanismos para permitir que os utilizadores consigam ignorar as secções de navegação, acelerando assim a navegação com o teclado.

Reviewer: ```suggestion
Providencia mecanismos para permitir que os utilizadores consigam ignorar as secções de navegação, acelerando assim a navegação com o teclado.
```

Skiplinks ou Skip Navigation Links são ligações escondidos que só ficam visíveis quando os utilizadores de teclado interagem com a página. São muito fáceis de implementar com alguns estilos e âncoras:

- [WebAIM - Skip Navigation Links](https://webaim.org/techniques/skipnav/)

> Use também elementos e pontos de referência como `<main>` e `<aside>`, para demarcar zonas da página, sendo que as tecnologias de acessibilidade permitem assim que o utilizador navegue rapidamente nessas secções.

Reviewer: ```suggestion
Usa também elementos e pontos de referência como `<main>` e `<aside>`, para demarcar zonas da página, sendo que as tecnologias de acessibilidade permitem assim que o utilizador navegue rapidamente nessas secções.
```

Saiba mais sobre o uso destes elementos para melhorar a acessibilidade aqui:

- [Accessible Landmarks](https://www.scottohara.me/blog/2018/03/03/landmarks.html)

### Controlar programaticamente o focus {#programmatically-managing-focus}

As nossas aplicações React modificam continuamente o HTML DOM durante o tempo de execução, às vezes perdendo o focus do teclado ou adicionado o focus a um elemento inesperado. Para corrigir isso, é necessário programar o focus do teclado na direção certo. Por exemplo, colocar o focus do teclado no botão que faz abrir uma modal depois dessa mesma modal ser fechada.

> Pode encontrar mais informações de como implementar no MDN Web Docs [keyboard-navigable JavaScript widgets](https://developer.mozilla.org/en-US/docs/Web/Accessibility/Keyboard-navigable_JavaScript_widgets).

Reviewer: ```suggestion
Podes encontrar mais informações de como implementar no MDN Web Docs [keyboard-navigable JavaScript widgets](https://developer.mozilla.org/en-US/docs/Web/Accessibility/Keyboard-navigable_JavaScript_widgets).
```

  }
}

// Agora podemos colocar o focus quando for necessário.
this.inputElement.current.focus();
```

> Ao usar um HOC para estender componentes, é recomendado [encaminhar a ref](/docs/forwarding-refs.html) para o elemento embrulhado usando a função de React `forwardRef`. se um third party HOC  não implementar o `forwardRef`, o padrão acima pode ser usado como fallback.

Reviewer: ```suggestion
Ao usar um HOC para estender componentes, é recomendado [encaminhar a ref](/docs/forwarding-refs.html) para o elemento embrulhado usando a função de React `forwardRef`. Se um HOC de terceiros não implementar o `forwardRef`, o padrão acima pode ser usado como fallback.
```

    }));
  }

  // Fechamos a popover no proximo tick usando o setTimeout.
  // É necessário porque precisamos primeiro de verificar o
  // outro filho do elemento que recebeu o focus

>   // visto que o evento blur foi acionado antes do novo evento de focus.

Author: creio que aqui será mesmo o termo "acionado" o mais correto.

>   // visto que o evento blur foi acionado antes do novo evento de focus.

Reviewer: boa! tens razão

}
```

Este código expõe a funcionalidade para os utilizadores com dispositivos de cursor (rato) e teclado. Observe também as props `aria-*` adicionadas para suportar utilizadores com leitores de ecrã. Por motivos de simplificação as interações das `arrow keys` nas opções da popover não foram implementadas. 

<img src="../images/docs/blur-popover-close.gif" alt="Uma lista popover corretamente fechada com o rato e com o teclado." />

Este é um exemplo de muitos casos em que depender apenas dos eventos de cursor, pode quebrar a funcionalidade para os utilizadores de teclado. Deve ir testando sempre com o teclado porque vai realçar logo as áreas problemáticas que podem ser corrigidas usando os eventos de reconhecimento de teclado.

## Widgets mais complexos {#more-complex-widgets}

> Uma experiência de utilização mais complexa não significa que seja menos acessível. Considerando que a acessibilidade é alcançada mais facilmente se programado o mais próximo possível do HTML, até mesmo o widget mais complexo pode ser programado de forma acessível.

Author: aqui tambem creio ser o mais correto o "se programado" em vez de "se programar-se"

> Uma experiência de utilização mais complexa não significa que seja menos acessível. Considerando que a acessibilidade é alcançada mais facilmente se programado o mais próximo possível do HTML, até mesmo o widget mais complexo pode ser programado de forma acessível.

Reviewer: concordo
